grudge/models/euler.py

```diff
@@ -373,4 +433,205 @@ def interp_to_quad_surf(u):
+    [DOI](https://doi.org/10.4208/cicp.170712.010313a)
+    """
+    dd_intfaces = tpair.dd
+    dd_allfaces = dd_intfaces.with_dtag("all_faces")
+    q_ll = tpair.int
+    q_rr = tpair.ext
+    actx = q_ll.array_context
+
+    num_flux = flux_chandrashekar(dcoll, q_ll, q_rr, gamma=gamma)
+    normal = thaw(dcoll.normal(dd_intfaces), actx)
+
+    if lf_stabilization:
+        from arraycontext import outer
+
+        rho_ll, u_ll, p_ll = conservative_to_primitive_vars(q_ll, gamma=gamma)
+        rho_rr, u_rr, p_rr = conservative_to_primitive_vars(q_rr, gamma=gamma)
+
+        def compute_wavespeed(rho, u, p):
+            return (
+                actx.np.sqrt(np.dot(u, u)) + actx.np.sqrt(gamma * (p / rho))
+            )
+
+        # Compute jump penalization parameter
+        lam = actx.np.maximum(compute_wavespeed(rho_ll, u_ll, p_ll),
+                              compute_wavespeed(rho_rr, u_rr, p_rr))
+        num_flux -= lam*outer(tpair.diff, normal)/2
+
+    return op.project(dcoll, dd_intfaces, dd_allfaces, num_flux @ normal)
+
+
+class EntropyStableEulerOperator(EulerOperator):
+
+    def operator(self, t, q):
+        from grudge.projection import volume_quadrature_project
+        from grudge.interpolation import \
+            volume_and_surface_quadrature_interpolation
+
+        gamma = self.gamma
+        qtag = self.qtag
+        dq = DOFDesc("vol", qtag)
+        df = DOFDesc("all_faces", qtag)
+
+        dcoll = self.dcoll
+
+        # Interpolate cv_state to vol quad grid: u_q = V_q u
+        q_quad = op.project(dcoll, "vol", dq, q)
+
+        # Convert to projected entropy variables: v_q = V_h P_q v(u_q)
+        entropy_vars = conservative_to_entropy_vars(q_quad, gamma=gamma)
+        proj_entropy_vars = volume_quadrature_project(dcoll, dq, entropy_vars)
+
+        # Compute conserved state in terms of the (interpolated)
+        # projected entropy variables on the quad grid
+        qtilde_allquad = \
+            entropy_to_conservative_vars(
+                # Interpolate projected entropy variables to
+                # volume + surface quadrature grids
+                volume_and_surface_quadrature_interpolation(
+                    dcoll, dq, df, proj_entropy_vars
+                ),
+                gamma=gamma)
+
+        # Compute volume derivatives using flux differencing
+        from functools import partial
+        from grudge.flux_differencing import volume_flux_differencing
+
+        def _reshape(shape, ary):
+            if not isinstance(ary, DOFArray):
+                return map_array_container(partial(_reshape, shape), ary)
+
+            return DOFArray(ary.array_context, data=tuple(
+                subary.reshape(grp.nelements, *shape)
+                # Just need group for determining the number of elements
+                for grp, subary in zip(dcoll.discr_from_dd("vol").groups, ary)))
+
+        flux_matrices = flux_chandrashekar(dcoll,
```

Comment:
```suggestion
        # FIXME These matrices are actually symmetric. Could make use
        # of that to avoid redundant computation.
        flux_matrices = flux_chandrashekar(dcoll,
```

Reply:
Done: https://github.com/inducer/grudge/pull/214/commits/927da0107dc1559757c37140b39d1a5dbe3e24d8